### PR TITLE
fix(tf2xla): propagate NaNs in min/max ops

### LINKED ISSUE #116992: CPU XLA silently drops NaN in `reduce_max` / `reduce_min` / `maximum` / `minimum`
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

 TensorFlow 2.21.0 stable

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

## Summary

When compiled with `tf.function(jit_compile=True)` and forced to CPU, TensorFlow XLA changes NaN semantics for min/max operations:

- `tf.reduce_max([nan, 1, 3, 2])` returns `3.0` under CPU XLA, while eager and   NumPy return `nan`.
- `tf.reduce_min([nan, 1, 3, 2])` returns `1.0` under CPU XLA, while eager and  NumPy return `nan`.
- `tf.math.maximum(tf.constant(5.0), tf.constant(nan))` returns `5.0` under CPU  XLA, while eager returns `nan`.
- `tf.math.minimum(tf.constant(5.0), tf.constant(nan))` returns `5.0` under CPU XLA, while eager returns `nan`.

This is a TensorFlow XLA integration / lowering issue rather than an XLA core limitation: JAX on CPU correctly propagates NaN for the same max/min operations.

## Why this is a bug

TensorFlow eager mode propagates NaN for `reduce_max`, `reduce_min`, `maximum`, and `minimum`, matching NumPy behavior. TensorFlow previously fixed NaN propagation for these eager paths in issues such as tensorflow/tensorflow#13046 and tensorflow/tensorflow#12659. Compiling the same TensorFlow program with `jit_compile=True` should not silently change this observable behavior.

NaNs are also realistic inputs: they can arise from upstream operations such as `0/0`, `inf - inf`, invalid normalization, or unstable training. Dropping them can mask invalid values and change downstream control flow or indices.

## Environment

- TensorFlow 2.21.0 stable: reproduces
- TensorFlow 2.22.0-dev20260424 nightly: reproduces
- CPU XLA only
- GPU XLA propagates NaN for these simple reduce/max/min cases
- JAX 0.9.2 CPU behavior matches NumPy and propagates NaN

## Notes

There is a related but separate GPU issue for `max_pool2d` and `tensor_scatter_nd_{min,max}` NaN behavior. That issue concerns XLA GPU propagating NaN where TensorFlow eager GPU ignores it. This issue is the inverse for CPU XLA simple min/max operations: CPU XLA drops NaN where TensorFlow eager and NumPy propagate it.

### Standalone code to reproduce the issue

```shell
## Minimal reproduction


import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""  # force CPU only
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"

import numpy as np
import tensorflow as tf

print(tf.__version__)

vals = tf.constant([float("nan"), 1.0, 3.0, 2.0], dtype=tf.float32)

print("reduce_max")
print("  eager:", tf.reduce_max(vals).numpy())
print("  xla:  ", tf.function(lambda x: tf.reduce_max(x), jit_compile=True)(vals).numpy())
print("  numpy:", np.max(vals.numpy()))

print("reduce_min")
print("  eager:", tf.reduce_min(vals).numpy())
print("  xla:  ", tf.function(lambda x: tf.reduce_min(x), jit_compile=True)(vals).numpy())
print("  numpy:", np.min(vals.numpy()))

a = tf.constant(5.0, dtype=tf.float32)
b = tf.constant(float("nan"), dtype=tf.float32)

print("maximum")
print("  eager:", tf.math.maximum(a, b).numpy())
print("  xla:  ", tf.function(lambda x, y: tf.math.maximum(x, y), jit_compile=True)(a, b).numpy())

print("minimum")
print("  eager:", tf.math.minimum(a, b).numpy())
print("  xla:  ", tf.function(lambda x, y: tf.math.minimum(x, y), jit_compile=True)(a, b).numpy())


Observed on TensorFlow 2.21.0 stable and TensorFlow 2.22.0-dev20260424:


reduce_max
  eager: nan
  xla:   3.0
  numpy: nan
reduce_min
  eager: nan
  xla:   1.0
  numpy: nan
maximum
  eager: nan
  xla:   5.0
minimum
  eager: nan
  xla:   5.0


## Downstream impact

The issue can be amplified into a discrete wrong result. For example, NaN
detection through max/min can change a boolean predicate, and reductions feeding
index selection can silently choose a finite element instead of preserving the
invalid value.


data = tf.constant([1.0, float("nan"), 3.0, 2.0], dtype=tf.float32)

def has_nan_via_max(x):
    return tf.math.is_nan(tf.reduce_max(x))

print(has_nan_via_max(data).numpy())  # True
print(tf.function(has_nan_via_max, jit_compile=True)(data).numpy())  # False on CPU XLA
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by anmolxlight**
Fixes #116992

Summary:
- propagate NaNs explicitly for XLA Maximum/Minimum kernels
- propagate NaNs explicitly for XLA ReduceMin/ReduceMax reducer bodies
- add regression tests for binary maximum/minimum and reduce min/max NaN behavior

Test plan:
- git diff --check
- python3 -m py_compile tensorflow/compiler/tests/binary_ops_test.py tensorflow/compiler/tests/reduce_ops_test.py
- bazel tests not run locally: bazel is not installed in this environment

### CODE REVIEW

File: tensorflow/compiler/tests/binary_ops_test.py [hunk 1]
```diff
@@ -477,6 +477,19 @@ def testNumericOps(self):
...
+      return
+    dtype = np.float32
+    values = np.array([[np.nan, 1.0], [5.0, np.nan]], dtype=dtype)
+    finite = np.array([2.0, 3.0], dtype=dtype)
+    maximum_expected = np.array([[np.nan, 3.0], [5.0, np.nan]], dtype=dtype)
+    minimum_expected = np.array([[np.nan, 1.0], [2.0, np.nan]], dtype=dtype)
+    self._testBinary(math_ops.maximum, values, finite, expected=maximum_expected)
+    self._testBinary(math_ops.maximum, finite, values, expected=maximum_expected)
+    self._testBinary(math_ops.minimum, values, finite, expected=minimum_expected)
+    self._testBinary(math_ops.minimum, finite, values, expected=minimum_expected)
```
[the remark this answers is not in the document]
Comment (anmolxlight) [reply tensorflow/compiler/tests/binary_ops_test.py]: Thanks for the review! Fixed — the test now iterates over `self.float_types` so NaN propagation is verified across float16, bfloat16, float64, and all other supported float types.

File: tensorflow/compiler/tests/reduce_ops_test.py [hunk 2]
```diff
@@ -145,6 +145,23 @@ def reference_max(dtype, inp, axis):
...
+        min_out = math_ops.reduce_min(a, index)
+        max_out = math_ops.reduce_max(a, index)
+
+      min_result, max_result = sess.run(
+          [min_out, max_out], {a: test_input, index: [0]})
+
+      self.assertTrue(np.isnan(min_result[0]))
+      self.assertEqual(min_result[1], 3.0)
+      self.assertTrue(np.isnan(max_result[0]))
+      self.assertEqual(max_result[1], 4.0)
```
[the remark this answers is not in the document]
Comment (anmolxlight) [reply tensorflow/compiler/tests/reduce_ops_test.py]: Good catch. Fixed — now loops over `self.float_types` and uses `self.assertAllClose` instead of `self.assertEqual` for more robust float comparison across precisions.

**Comment by anmolxlight** [comment]
@sdasgup3 @gbaned — addressed the gemini-code-assist[bot] review feedback:

**binary_ops_test.py** — `testMaximumMinimumPropagateNan` now iterates over `self.float_types` instead of only `np.float32`, verifying NaN propagation across float16, bfloat16, float64, and all other supported float types.

**reduce_ops_test.py** — `testReduceMinMaxPropagateNan` similarly loops over `self.float_types` and uses `assertAllClose` instead of `assertEqual` for more robust floating-point comparison across precisions.

Both threads resolved. Ready for re-review.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for addressing the review feedback and expanding the tests (`testMaximumMinimumPropagateNan` in `binary_ops_test.py` and `testReduceMinMaxPropagateNan` in `reduce_ops_test.py`) to cover all float precisions (`float16`, `bfloat16`, and `float64`). The coverage is highly valuable.
  
  However, we have a critical performance concern regarding the reduction operations in `reduction_ops.cc` before this can be merged.
  
  ### TPU/GPU Code Vectorization Issues
  By inserting nested `Select` and `IsNan` checking in the combiner body of `ReduceMin` and `ReduceMax` (inside `BuildReducer`):
  ```cpp
  return xla::Select(xla::IsNan(scalar_lhs), scalar_lhs,
                     xla::Select(xla::IsNan(scalar_rhs), scalar_rhs, result));
  ```
We alter the basic pattern of the reduction. During compile-time (particularly for GPUTPU backends), the compiler expects standard, simple combiners (like xla::Max/xla::Min primitives). When it encounters custom combiners with conditional selections, it fails to map the reduction to hardware vector reduction instructions. This forces the XLA compiler to emit unvectorized elementwise reduction loops, introducing a severe performance degradation for standard tf.reduce_min and tf.reduce_max ops.

**Proposed Alternatives**
**Target-Level Compiler Flags:** Have you verified if the native XLA compiler flags (such as setting --xla_gpu_enable_fast_min_max=false) already propagate NaNs correctly under certain configurations? Handing this at the backend/backend-config level is generally preferred to preserve optimizations.

**Pre-filtering / Fast path: I**f we must solve this in the tf2xla bridge, can we check for NaNs on the input tensor outside the combiner logic (e.g. running a fast check/pass to see if NaNs are present, and only falling back to the slow loops if NaNs exist)? This leaves the fast-path path vectorized.  

Please also fix PyLint check issues.
